adapters/adhese/adhese.go

```
@@ -0,0 +1,247 @@
+}
+
+func convertAdheseBid(adheseBid AdheseBid, adheseExt AdheseExt, adheseOriginData AdheseOriginData) openrtb.BidResponse {
+	adheseExtJson, _ := json.Marshal(adheseOriginData)
+
+	return openrtb.BidResponse{
+		ID: adheseExt.Id,
+		SeatBid: []openrtb.SeatBid{{
+			Bid: []openrtb.Bid{{
+				DealID: adheseExt.OrderId,
+				CrID:   adheseExt.Id,
+				AdM:    getAdMarkup(adheseBid, adheseExt),
+				Ext:    adheseExtJson,
+			}},
+			Seat: "",
+		}},
+	}
+}
+
+func convertAdheseOpenRtbBid(adheseBid AdheseBid) openrtb.BidResponse {
+	var response openrtb.BidResponse = adheseBid.OriginData
+	if len(response.SeatBid) > 0 && len(response.SeatBid[0].Bid) > 0 {
+		response.SeatBid[0].Bid[0].AdM = adheseBid.Body
+	}
+	return response
+}
+
+func getAdMarkup(adheseBid AdheseBid, adheseExt AdheseExt) string {
+	if adheseExt.Ext == "js" {
+		if ContainsAny(adheseBid.Body, []string{"<script", "<div", "<html"}) {
+			return adheseBid.Body + "<img src='" + adheseExt.ImpressionCounter + "' style='height:1px; width:1px; margin: -1px -1px; display:none;'/>"
+		}
+		if ContainsAny(adheseBid.Body, []string{"<?xml", "<vast"}) {
+			return adheseBid.Body
+		}
+	}
+	return adheseExt.Tag
+}
+
+func getBidType(bidAdm string) openrtb_ext.BidType {
+	if bidAdm != "" && ContainsAny(bidAdm, []string{"<?xml", "<vast"}) {
+		return openrtb_ext.BidTypeVideo
+	}
+	return openrtb_ext.BidTypeBanner
+}
+
+func WrapError(errorStr string) *errortypes.BadInput {
+	return &errortypes.BadInput{Message: errorStr}
+}
+
+func ContainsAny(raw string, keys []string) bool {
+	for i := 0; i < len(keys); i++ {
+		if strings.Contains(strings.ToLower(raw), keys[i]) {
+			return true
+		}
+	}
+	return false
+
+}
+
+func NewAdheseAdapter(config *adapters.HTTPAdapterConfig, uri string) *AdheseAdapter {
+	return NewAdheseBidder(adapters.NewHTTPAdapter(config).Client, uri)
+}
+
+func NewAdheseBidder(client *http.Client, uri string) *AdheseAdapter {
+	template, _ := template.New("endpointTemplate").Parse(uri)
```

Please do not ignore errors.